spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Adam.scala

```diff
@@ -64,22 +75,49 @@ class Adam[@specialized(Float, Double) T: ClassTag](
     val (_s, _r, _denom) =
       if (state.get[Tensor[T]]("s").isDefined) {
         (state.get[Tensor[T]]("s").get, state.get[Tensor[T]]("r").get,
-          Tensor[T]().resizeAs(dfdx).zero())
+          buffer3.resizeAs(dfdx))
       } else {
-        (Tensor[T]().resizeAs(dfdx).zero(), Tensor[T]().resizeAs(dfdx).zero(),
-          Tensor[T]().resizeAs(dfdx).zero())
+        (buffer1.resizeAs(dfdx).zero(), buffer2.resizeAs(dfdx).zero(),
+          buffer3.resizeAs(dfdx))
       }
+
+    // used as MKL.axpy: 1 * a + y = y
+    if (buffer5.nElement < dfdx.nElement) {
+      buffer5.resizeAs(dfdx).fill(ev.one)
+    }
+
     val clr = lr / (1 + timestep*lrd)
 
     timestep = timestep + 1
 
+    /**
+     * m_t = beta_1 * m_t-1 + (1 - beta_1) * g_t
+     * v_t = beta_2 * v_t-1 + (1 - beta_2) * g_t * g_t
+     */
     _s.mul(ev.fromType[Double](beta1)).add(ev.fromType[Double](1-beta1), dfdx)
-    _r.mul(ev.fromType[Double](beta2)).addcmul(ev.fromType[Double](1-beta2), dfdx, dfdx)
-    _denom.resizeAs(_r).copy(_r).sqrt().add(ev.fromType[Double](eps))
+    // buffer4 = dfdx * dfdx
+    buffer4.resizeAs(dfdx).cmul(dfdx, dfdx)
+
+    _r.mul(ev.fromType[Double](beta2)).add(ev.fromType[Double](1-beta2), buffer4)
+    ev.vSqrt(_r.nElement,
```

Comment:
_r and _denom is contiguous, why not just use _denom.sqrt(_r)

Reply:
ok, changed